Mathlib/Algebra/Module/Presentation/Basic.lean

```diff
@@ -0,0 +1,328 @@
+
+lemma surjective_mapToKer : Function.Surjective solution.mapToKer := by
+  rintro ⟨x, hx⟩
+  rw [h.ker_π, ← relations.range_map] at hx
+  obtain ⟨r, rfl⟩ := hx
+  exact ⟨r, rfl⟩
+
+/-- The sequence `(relations.R →₀ A) → (relations.G →₀ A) → M → 0` is exact. -/
+lemma exact : Function.Exact relations.map solution.π := by
+  intro x₂
+  constructor
+  · intro hx₂
+    obtain ⟨x₁, hx₁⟩ := h.surjective_mapToKer ⟨x₂, hx₂⟩
+    exact ⟨x₁, by simpa only [mapToKer_coe, Subtype.ext_iff] using hx₁⟩
+  · rintro ⟨x₁, rfl⟩
+    rw [π_comp_map_apply]
+
+variable {N : Type v'} [AddCommGroup N] [Module A N]
+
+/-- If `M` admits a presentation by generators and relations, and we have a solution of the
+same equations in a module `N`, then this is the canonical induced linear map `M →ₗ[A] N`. -/
+noncomputable def desc (s : relations.Solution N) : M →ₗ[A] N :=
+  s.fromQuotient.comp h.linearEquiv.symm.toLinearMap
+
+@[simp]
+lemma desc_var (s : relations.Solution N) (g : relations.G) :
+    h.desc s (solution.var g) = s.var g := by
+  dsimp [desc]
+  rw [linearEquiv_symm_var, fromQuotient_mk, π_single]
+
+lemma postcomp_injective {f f' : M →ₗ[A] N}
+    (h' : solution.postcomp f = solution.postcomp f') : f = f' := by
+  suffices f.comp solution.fromQuotient = f'.comp solution.fromQuotient by
+    ext x
+    obtain ⟨y, rfl⟩ := h.bijective.2 x
+    exact DFunLike.congr_fun this y
+  ext g
+  simpa using congr_fun (congr_arg Solution.var h') g
+
+/-- If `M` admits a presentation by generators and relations, then
+linear maps from `M` can be (naturally) identified to the solutions of
+certain linear equations. -/
+@[simps]
+noncomputable def linearMapEquiv : (M →ₗ[A] N) ≃ relations.Solution N where
+  toFun f := solution.postcomp f
+  invFun s := h.desc s
+  left_inv f := h.postcomp_injective (by aesop)
+  right_inv s := by aesop
+
+end IsPresentation
+
+variable (relations)
+
+/-- Given `relations : Relations A`, this is the obvious solution to `relations`
+in the quotient `relations.Quotient`. -/
+noncomputable def ofQuotient : relations.Solution relations.Quotient :=
+  ofπ relations.toQuotient (by simp)
+
+@[simp]
+lemma ofQuotient_π : (ofQuotient relations).π = Submodule.mkQ _ := ofπ_π _ _
+
+@[simp]
+lemma ofQuotient_fromQuotient : (ofQuotient relations).fromQuotient = .id := by aesop
+
+lemma ofQuotient_isPresentation : (ofQuotient relations).IsPresentation where
+  bijective := by
+    simpa only [ofQuotient_fromQuotient, LinearMap.id_coe] using Function.bijective_id
+
+end Solution
+
+end Relations
+
+variable (M : Type v) [AddCommGroup M] [Module A M]
+
+/-- Given an `A`-module `M`, a term in this type is a presentation by `M` by
+generators and relations. -/
+@[nolint checkUnivs]
+structure Presentation where
+  /-- the shape of the relations -/
+  relations : Relations A
+  /-- a solution to the given equations -/
+  solution : relations.Solution M
+  isPresentation : solution.IsPresentation
```

Comment:
If you implement it via
```lean
/-- Given an `A`-module `M`, a term in this type is a presentation by `M` by
generators and relations. -/
@[nolint checkUnivs]
structure Presentation extends Relations A, toRelations.Solution M, toSolution.IsPresentation where
```
Then you can directly write
```
example : Presentation A M where
  G := _
  R := _
  relation := _
  var := _
  relation := _
  bijective := _
```
when defining one.
(There are two fields named `relation` so you might need to rename one)

I have no preferences either way though. It's up to you.

Reply:
Thanks for the suggestion. I have expanded the name of `Solution.relation`.